### PR TITLE
Add roles

### PR DESCRIPTION
**Pull request by YuriyLu**

### CODE REVIEW

File: contracts/Peeranha.sol [hunk 1]
```diff
@@ -38,8 +40,34 @@ contract Peeranha is IPeeranha, Initializable, AccessControlUpgradeable, ERC20Up
...
+
+    modifier onlyCommunityModerator(uint256 communityId) {
+        require((hasRole(bytes32(COMMUNITY_MODERATOR_ROLE + communityId), msg.sender) || 
+            hasRole(DEFAULT_ADMIN_ROLE, msg.sender)), 
+            "Peeranha: must have community moderator role");
+        _;
+    }
+
+    modifier onlyExisitingCommunity(uint256 communityId) {
+        require(communities.getCommunitiesCount() > communityId, "Peeranha: must be an existing community");
```
Comment (sergeyi23) [inline contracts/Peeranha.sol]: This modifier should check that community is not frozen. Did you plan another modifier for that?

File: contracts/Peeranha.sol [hunk 2]
```diff
@@ -38,8 +40,34 @@ contract Peeranha is IPeeranha, Initializable, AccessControlUpgradeable, ERC20Up
...
+        _;
+    }
 
-    function __Peeranha_init() internal initializer {
+    modifier onlyExisitingUser(address user) {
+        users.getUserByAddress(user);
```
Comment (sergeyi23) [inline contracts/Peeranha.sol]: Add another function to UserLib that checks uses existence. `getUserByAddress` reads to much data that we do not use. This is wasted gas.

File: contracts/Peeranha.sol [hunk 3]
```diff
@@ -118,11 +148,95 @@ contract Peeranha is IPeeranha, Initializable, AccessControlUpgradeable, ERC20Up
      * Requirements:
      *
      * - Must be an existing community.  
+     * - Sender must be community moderator.
      */
-    function updateCommunity(uint256 communityId, bytes32 ipfsHash) external {
+    function updateCommunity(uint256 communityId, bytes32 ipfsHash) external onlyCommunityModerator(communityId) {
```
Comment (sergeyi23) [inline contracts/Peeranha.sol]: I think only admin should be allowed to edit the community. Moderators will be deleting questions, etc.

File: contracts/Peeranha.sol [hunk 4]
```diff
@@ -118,11 +148,95 @@ contract Peeranha is IPeeranha, Initializable, AccessControlUpgradeable, ERC20Up
...
 
+    /**
+     * @dev Freeze community.
+     *
+     * Requirements:
+     *
+     * - Must be an existing community.  
+     * - Sender must be community moderator.
+     */
+    function freezeCommunity(uint256 communityId) external onlyCommunityModerator(communityId) {
```
Comment (sergeyi23) [inline contracts/Peeranha.sol]: Also only admin should be allowed to execute this action.

Comment (sergeyi23) [reply contracts/Peeranha.sol]: Global admin also should be allowed to freeze any community

File: contracts/Peeranha.sol [hunk 5]
```diff
@@ -118,11 +148,95 @@ contract Peeranha is IPeeranha, Initializable, AccessControlUpgradeable, ERC20Up
...
+    /**
+     * @dev Give community adminisrator permission.
+     *
+     * Requirements:
+     *
+     * - Sender must be global administrator.
+     * - Must be an existing community.
+     * - Must be an existing user. 
+     */
+    function giveCommunityAdminPermission(address user, uint256 communityId) external
```
Comment (sergeyi23) [inline contracts/Peeranha.sol]: I would prefer to use `_grantRole` here and allow any global admin to grant community admin permissions

File: contracts/Peeranha.sol [hunk 6]
```diff
@@ -118,11 +148,95 @@ contract Peeranha is IPeeranha, Initializable, AccessControlUpgradeable, ERC20Up
...
+    /**
+     * @dev Revoke community adminisrator permission.
+     *
+     * Requirements:
+     *
+     * - Sender must be global administrator.
+     * - Must be an existing community.
+     * - Must be an existing user. 
+     */
+    function revokeCommunityAdminPermission(address user, uint256 communityId) external
```
Comment (sergeyi23) [inline contracts/Peeranha.sol]: Similar here. Use `_revokeRole` and allow any global admin to execute.

File: contracts/Peeranha.sol [hunk 7]
```diff
@@ -38,8 +40,34 @@ contract Peeranha is IPeeranha, Initializable, AccessControlUpgradeable, ERC20Up
...
+
+    modifier onlyCommunityAdmin(uint256 communityId) {
+        require((hasRole(bytes32(COMMUNITY_ADMIN_ROLE + communityId), msg.sender) || 
+            hasRole(DEFAULT_ADMIN_ROLE, msg.sender)), 
+            "Peeranha: must have community admin role");
+        _;
+    }
+
+    modifier onlyCommunityModerator(uint256 communityId) {
+        require((hasRole(bytes32(COMMUNITY_MODERATOR_ROLE + communityId), msg.sender) ||
```
Comment (sergeyi23) [inline contracts/Peeranha.sol]: Move expression `bytes32(COMMUNITY_MODERATOR_ROLE + communityId)` to a function and reuse.

File: contracts/libraries/CommunityLib.sol [hunk 8]
```diff
@@ -20,16 +20,30 @@ library CommunityLib {
...
         uint8 tagsCount;
+        bool isFrozen;
     }
 
     struct CommunityCollection {
         mapping(uint256 => CommunityContainer) communities;
         uint8 communityCount;
     }
 
+    modifier onlyExistedAndNotFrozen(CommunityCollection storage self, uint256 id){
```
Comment (sergeyi23) [inline contracts/libraries/CommunityLib.sol]: rename to `onlyExistingAndNotFrozen`

File: contracts/libraries/CommunityLib.sol [hunk 9]
```diff
@@ -20,16 +20,30 @@ library CommunityLib {
...
+        require(!self.communities[id].isFrozen,
+            "Community is frozen"
+        );
+        _;
+    }
+
     event CommunityCreated(uint256 id, bytes32 ipfsHash, bytes32 ipfsHash2, Tag[] tags);
     event CommunityUpdated(uint256 id, bytes32 ipfsHash);
     event TagCreated(uint256 communityId, uint256 tagId, bytes32 ipfsHash, bytes32 ipfsHash2);
+    event Freeze(uint256 commintyId);
```
Comment (sergeyi23) [inline contracts/libraries/CommunityLib.sol]: Let's do longer name `CommunityFrozen` and `CommunituUnfrozen`

**Review by sergeyi23 — CHANGES_REQUESTED** [review]
See comments

File: contracts/Peeranha.sol (hunk 3 above)
Comment (YuriyLu) [reply contracts/Peeranha.sol]: You mean _setupRole?

File: contracts/Peeranha.sol (hunk 5 above)
Comment (YuriyLu) [reply contracts/Peeranha.sol]: You mean _setupRole?

Comment (YuriyLu) [reply contracts/Peeranha.sol]: Global admin is DEFAULT_ADMIN_ROLE, he already has all roles and can do this

**Pull request merged**